.github/workflows/retry.yml

```diff
@@ -0,0 +1,184 @@
+# Please ping @DilumAluthge when making any changes to this file.
+
+# Here are some steps that we take in this workflow file for security reasons:
+# 1. We do not checkout any code.
+# 2. We do not run any external actions.
+# 3. We do not give the `GITHUB_TOKEN` any permissions.
+# 4. We only give the Buildkite API token (`BUILDKITE_API_TOKEN_RETRY`) the minimum necessary
+#    set of permissions.
+
+# Important note to Buildkite maintainers:
+# In order to make this work, you need to tell Buildkite that it should NOT create a brand-new
+# build when someone closes and reopens a pull request. To do so:
+# 1. Go to the relevant pipeline (e.g. https://buildkite.com/julialang/julia-master).
+# 2. Click on the "Pipeline Settings" button.
+# 3. In the left sidebar, under "Pipeline Settings", click on "GitHub".
+# 4. In the "GitHub Settings", under "Build Pull Requests", make sure that the "Skip pull
+#    request builds for existing commits" checkbox is checked. This is the setting that tells
+#    Buildkite that it should NOT create a brand-new build when someone closes and reopens a
+#    pull request.
+# 5. At the bottom of the page, click the "Save GitHub Settings" button.
+
+name: Retry Failed Buildkite Jobs
+
+on:
+  # When using the `pull_request_target` event, all PRs will get access to secret environment
+  # variables (such as the `BUILDKITE_API_TOKEN_RETRY` secret environment variable), even if
+  # the PR is from a fork. Therefore, for security reasons, we do not checkout any code in
+  # this workflow.
+  # TODO: change `pull_request` to `pull_request_target`.
+  pull_request:
+
+    # TODO: delete the following line (once we have completely transitioned from Buildbot to Buildkite)
+    types: [ reopened, labeled ]
+
+    # TODO: uncomment the following line (once we have completely transitioned from Buildbot to Buildkite)
+    # types: [ reopened ]
```

Comment:
Are these TODOs still relevant?

Reply:
This deserves some explanation.

Okay, so if you look at the workflow file currently, you'll see it runs on two types of event:
1. When the `Buildkite - retry failed jobs` label is added to a PR.
2. When a PR is reopened after being closed.

Once we have fully migrated to Buildkite, I will remove option 1 and only have option 2.

In the current situation, in which we have both Buildbot and Buildkite, here is what happens in each event:
1. Label is added: In this event, all of the failed Buildkite jobs will be rerun. None of the passed Buildkite jobs will be rerun. **Nothing happens to the Buildbot jobs.**
2. PR is reopened after being closed: In this event, all of the failed Buildkite jobs will be rerun. None of the passed Buildkite jobs will be rerun. **All of the Buildbot jobs (both passed and failed) will be rerun.**

Once we have migrated everything to Buildkite, and we no longer have anything on Buildbot, here is what happens in each event:
1. Label is added: In this event, all of the failed Buildkite jobs will be rerun.
2. PR is reopened after being closed: In this event, all of the failed Buildkite jobs will be rerun.

So, currently (in the Buildbot + Buildkite world), the "Label is added" event is preferable, because it doesn't rerun the passed Buildbot jobs, and thus conserves CI resources.

But in the future Buildkite-only world, the "Label is added" event is exactly identical in effect to the "PR is reopened after being closed" event. And I think, personally, that telling someone to "close and reopen a PR, and it will only rerun failed CI" is a much more user-friendly interface than this label business.

So, once we are in the Buildkite-only world, I want to completely remove the label trigger, and only have the "close and reopen the PR" trigger, because I think it's easier for people to use.

So these TODOs are here to remind me to get rid of the label stuff once we are Buildkite-only.

What do you think?